### PR TITLE
fix: live reload port fallback if port is used

### LINKED ISSUE #731: Try new Live Reload ports if default is taken
### Is this a bug report?

Not really. Sort of.

### Have you read the [Contributing Guidelines on issues](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md#reporting-new-issues)?

Yes

### Environment

N/A

### Steps to Reproduce

1. Run yarn start on a site
2. Run yarn start on another site

### Expected Behavior

Both sites run.

### Actual Behavior

The server port will get to 3001. But the Live Reload port errors out as duplicate.

### Reproducible Demo

See steps to reproduce above

### PR DESCRIPTION
**Pull request by tom-auger**
## Motivation

Fixes #731

1. Factored out the checking of an unused port to a separate module and use it when starting the Docusaurus server and the Live Reload server. 
1. Refactored the starting of servers into a separate module to enable testing.
1. If the Live Reload server fails to start the Docusaurus server will still start but without live reload functionality.
1. Added Jest tests for new modules.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

1. Added Jest tests around new code and checked they passed.
1. Manually started multiple instances and checked the servers started on separate ports and the sites were accessible.

## Related PRs

N/A

### CODE REVIEW

File: lib/__tests__/findUnusedPort.test.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
+/**
```
Comment (JoelMarcey) [inline lib/__tests__/findUnusedPort.test.js]: ❤️ that you are adding tests.

File: lib/server/__mocks__/tiny-lr.js [hunk 2]
```diff
@@ -5,6 +5,8 @@
  * LICENSE file in the root directory of this source tree.
  */
 
-module.exports = {
-  LIVE_RELOAD_PORT: 35729,
+const tinylrServer = {
+  listen: jest.fn(),
```
Comment (JoelMarcey) [inline lib/server/__mocks__/tiny-lr.js]: This file has no relationship to the `constants` file, right? You just changed the `constants` file and then made a file move. I am just wondering why GitHub thinks this is a file move.

Comment (tom-auger) [reply lib/server/__mocks__/tiny-lr.js]: That's right, the two files are unrelated. I actually deleted the `constants` file in a separate commit, and added the `tiny-lr` mock in a later one, so I'm not sure why GitHub thinks this was a file move 😕

**Comment by tom-auger** [comment]
@yangshun thanks 😃, sure I can fix it up! RE the comments by @endiliey:

1. The node-portfinder package looks simple to drop in as a replacement to my `findUnusedPort` implementation. I'm happy to do that if you think it's better?
1. Will rename and move `serverController.js` as suggested.
1. Do you want to keep any of the tests I've added? I don't mind removing some or all of them if you don't need them 😃

**Comment by tom-auger** [comment]
@yangshun I've made the suggested changes. I fixed the tests too, but if you don't want to keep them feel free to remove them 😄

**Pull request merged**